File: Python3.x/338-Counting Bits.py
# Runtime: 132 ms, faster than 52.84% of Python3 online submissions for Counting Bits.
# Memory Usage: 15.9 MB, less than 5.22% of Python3 online submissions for Counting Bits.
class Solution:
    def countBits(self, num: int) -> List[int]:
        result = [0]
        for i in range(1, num + 1):
            result.append(result[i >> 1] + (i & 1))
        return result


'''
0 (0, '0b0')
1 (1, '0b1')
1 (2, '0b10')
2 (3, '0b11')
1 (4, '0b100')
2 (5, '0b101')
2 (6, '0b110')
3 (7, '0b111')
1 (8, '0b1000')
2 (9, '0b1001')

find regular
'''


# Runtime: 104 ms, faster than 97.24% of Python3 online submissions for Counting Bits.
# Memory Usage: 15.8 MB, less than 68.15% of Python3 online submissions for Counting Bits.
class Solution:
    def countBits(self, num: int) -> List[int]:
        res = [0]
        while len(res) <= num:
            res += [x + 1 for x in res]
        return res[:num + 1]

File: Algorithms/C++/338-Counting_Bits.cpp
// Runtime: 52 ms, faster than 91.45% of C++ online submissions for Counting Bits.
// Memory Usage: 9.4 MB, less than 100.00% of C++ online submissions for Counting Bits.
class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> res(num + 1, 0);
        for (int i = 1; i <= num; ++i) {
            int tmp = i, n = 0;
            while (tmp > 0) {
                if (tmp & 1 == 1)
                    ++n;
                tmp >>= 1;
            }
            res[i] = n;
        }
        return res;
    }
};


// Runtime: 52 ms, faster than 91.45% of C++ online submissions for Counting Bits.
// Memory Usage: 9.4 MB, less than 100.00% of C++ online submissions for Counting Bits.
class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> res(num + 1, 0);
        for (int i = 1; i <= num; ++i)
            res[i] = res[i & (i - 1)] + 1;
        return res;
    }
};


// Runtime: 52 ms, faster than 91.45% of C++ online submissions for Counting Bits.
// Memory Usage: 9.6 MB, less than 81.70% of C++ online submissions for Counting Bits.
class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> res(num + 1, 0);
        for (int i = 1; i <= num; ++i)
            res[i] = res[i >> 1] + (i & 1);  // Recursive formula
        return res;
    }
};
